packages/jspsych/src/modules/plugin-api/KeyboardListenerAPI.ts

```diff
@@ -1,176 +1,149 @@
+  callback_function: any;
+  valid_responses?: ValidResponses;
+  rt_method?: "performance" | "audio";
+  persist?: boolean;
+  audio_context?: AudioContext;
+  audio_context_start_time?: number;
+  allow_held_key?: boolean;
+  /**
+   * overiding via parameters for testing purposes.
+   * TODO (bjoluc): Why exactly?
+   */
+  minimum_valid_rt?: number;
+}
 
-  private keyboard_listeners = [];
+export class KeyboardListenerAPI {
+  constructor(private areResponsesCaseSensitive: boolean = false, private minimumValidRt = 0) {}
 
-  private held_keys = {};
+  private listeners = new Set<KeyboardListener>();
+  private heldKeys = new Set<string>();
 
-  private root_keydown_listener(e) {
-    for (var i = 0; i < this.keyboard_listeners.length; i++) {
-      this.keyboard_listeners[i].fn(e);
+  private rootKeydownListener(e: KeyboardEvent) {
+    // Iterate over a static copy of the listeners set because listeners might add other listeners
+    // that we do not want to be included in the loop
+    for (const listener of Array.from(this.listeners)) {
+      listener.fn(e);
     }
-    this.held_keys[e.key] = true;
+    this.heldKeys.add(this.toLowerCaseIfInsensitive(e.key));
   }
 
-  private root_keyup_listener(e) {
-    this.held_keys[e.key] = false;
+  private toLowerCaseIfInsensitive(string: string) {
+    return this.areResponsesCaseSensitive ? string : string.toLowerCase();
   }
 
-  reset(root_element) {
-    this.keyboard_listeners = [];
-    this.held_keys = {};
-    root_element.removeEventListener("keydown", this.root_keydown_listener);
-    root_element.removeEventListener("keyup", this.root_keyup_listener);
+  private rootKeyupListener(e: KeyboardEvent) {
+    this.heldKeys.delete(this.toLowerCaseIfInsensitive(e.key));
   }
 
-  createKeyboardEventListeners(root_element) {
-    root_element.addEventListener("keydown", this.root_keydown_listener);
-    root_element.addEventListener("keyup", this.root_keyup_listener);
+  private isResponseValid(validResponses: ValidResponses, allowHeldKey: boolean, key: string) {
+    // check if key was already held down
+    if (!allowHeldKey && this.heldKeys.has(key)) {
+      return false;
+    }
+
+    if (validResponses === "ALL_KEYS") {
+      return true;
+    }
+    if (validResponses === "NO_KEYS") {
+      return false;
+    }
+
+    return validResponses.includes(key);
   }
 
-  getKeyboardResponse(parameters) {
-    //parameters are: callback_function, valid_responses, rt_method, persist, audio_context, audio_context_start_time, allow_held_key
+  reset(root_element: Element) {
+    this.listeners.clear();
+    this.heldKeys.clear();
+    root_element.removeEventListener("keydown", this.rootKeydownListener);
+    root_element.removeEventListener("keyup", this.rootKeyupListener);
+  }
+
+  createKeyboardEventListeners(root_element: Element) {
+    root_element.addEventListener("keydown", this.rootKeydownListener);
+    root_element.addEventListener("keyup", this.rootKeyupListener);
+  }
```

Comment:
@jodeleeuw These methods are not mentioned in the docs. Any reason for them to be publicly available? I can't think of a scenario in which users would need to call these. Also, calling `reset()` in a plugin without calling `createKeyboardEventListeners()` afterwards will break keyboard responses in subsequent trials. I think the root element should be passed to the constructor instead which sets everything up then.

Reply:
No, these were meant to be internal methods. I'm fine with moving to a constructor.

Reply:
Few notes on how I implemented this now:
* Only in the constructor wouldn't have worked since the JsPsych `DOM_container` is currently only known when `run()` is called
* I didn't want to leave a public setup function, so a function is injected into the constructor now that returns the display element, if known. For unit tests (and if we change the `DOM_container` setup in jsPsych), that function is already invoked in the constructor. If it did not return an element then, the listeners are registered on the first `getKeyboardResponse()` call when the display element is already known.
* I used this chance to remove the dependency on JsPsych and the html-keyboard-response plugin in `pluginapi.test.ts`.
* When we change the internal JsPsych setup behavior during some future refactoring, we can also get rid of the function injection approach without changing any public API :)

Reply:
Since the `display_element` parameter is passed to `initJsPsych` and not `run()` couldn't we know the value of `DOM_container` earlier? I suppose one reason for delaying it is if you wanted to dynamically setup a container between the time that you called `initJsPsych` and `run` but I can't think of a reason that you couldn't do the dynamic initialization prior to calling `initJsPsych`. 

There's no legacy reason to not move it to `initJsPsych` as far as I know, since in `v6.x` `run()` and `initJsPsych` were essentially the same thing.

Reply:
I'd love to move it, but I just went over the code again and found the blocker: `setupDOM()` is asynchronous because it waits for the document to be ready before it creates the container elements (#948). So there's no way to reliably know the container elements in the constructor...